uint32 x_obj
uint32 y_obj